actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,77 @@
+require 'active_support/core_ext/object/try'
+
+module ActionView
+  module CollectionCaching # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      # Fallback cache store if Action View is used without Rails.
+      # Otherwise overriden in Railtie to use Rails.cache.
+      mattr_accessor(:collection_cache) { ActiveSupport::Cache::MemoryStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }
+      rendered_partials = @collection.any? ? yield : []
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
```

Comment:
I was thinking about this some more. It's a shame we `Enumerator` raises a `StopIteration` exception, otherwise this would have worked (and we wouldn't have to dup the whole return value):

```ruby
rendered_partials = @collection.any? ? yield.each : []

fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
  rendered_partials.next
end
```

The weird thing was adding this didn't make any tests blow up, which kind of worries me.